Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift

```diff
@@ -201,17 +198,37 @@ struct HTTPRequestStateMachine {
             self.state = .failed(error)
             return .failRequest(error, .none)
 
-        case .running(.streaming, .receivingBody),
-             .running(.endSent, .receivingBody)
-                 where error as? NIOSSLError == .uncleanShutdown && self.ignoreUncleanSSLShutdown == true:
+        case .running(.streaming, .waitingForHead),
+             .running(.endSent, .waitingForHead) where error as? NIOSSLError == .uncleanShutdown:
+            // if we received a NIOSSL.uncleanShutdown before we got an answer we should handle
+            // this like a normal connection close. We will receive a call to channelInactive after
+            // this error.
             return .wait
 
+        case .running(.streaming, .receivingBody(let responseHead, _)),
+             .running(.endSent, .receivingBody(let responseHead, _)) where error as? NIOSSLError == .uncleanShutdown:
+
+            // if we have already received the response head, the parser will ensure that we receive
+            // the complete response. we can ignore this error. we might see a HTTPParserError very
+            // soon.
+            if responseHead.headers.contains(name: "content-length") || responseHead.headers.contains(name: "transfer-encoding") {
```

Comment:
This is not a sufficient check: there are cases where these headers are not present, but we are nonetheless confident of the length of the response. Specifically:

1. Responses to a HEAD request
2. Responses with a 204 or 304 status code

Reply:
The `HTTPResponseDecoder` will always emit a response `.end` after having received a response head, if the request had a `HEAD` or `CONNECT` method or the response status is 204 or 304. 

https://github.com/apple/swift-nio/blob/main/Sources/NIOHTTP1/HTTPDecoder.swift#L301-L314

For this reason, we can expect that we will only be in the streaming state, if we have a "content-length" or "transfer-encoding" header or if the response is EOF terminated. I'm well aware that this is _very_ implicit. If we want to have explicit checks here, I will need to come up with a way to preserve the request method a little longer.

Reply:
Hmm, can we get this logic into a comment then? The correctness of this code is not apparent merely from reading it.

Reply:
Comment added.